File: services/federation-product-exchange-service/src/domain/encryption.rs
use crate::models::{
    contract::SharingContract,
    share::SharedDataset,
    sync_status::{EncryptionPosture, SyncStatus},
};

pub fn posture(
    share: &SharedDataset,
    contract: Option<&SharingContract>,
    sync_status: Option<&SyncStatus>,
) -> EncryptionPosture {
    let profile = contract
        .map(|contract| contract.encryption_profile.clone())
        .unwrap_or_else(|| "mutual-tls+envelope".to_string());
    let key_version = sync_status
        .map(|status| status.key_version.clone())
        .unwrap_or_else(|| "key/pending".to_string());
    let encrypted_in_transit = sync_status
        .map(|status| status.encrypted_in_transit)
        .unwrap_or(true);
    let encrypted_at_rest = sync_status
        .map(|status| status.encrypted_at_rest)
        .unwrap_or(true);
    let recommendation = if encrypted_in_transit && encrypted_at_rest {
        "rotation current"
    } else {
        "provision key rotation before replication"
    };

    EncryptionPosture {
        share_id: share.id,
        transport_cipher: if encrypted_in_transit {
            "TLS 1.3 mTLS".to_string()
        } else {
            "disabled".to_string()
        },
        at_rest_cipher: if encrypted_at_rest {
            "AES-256-GCM envelope".to_string()
        } else {
            "disabled".to_string()
        },
        key_version,
        profile,
        encrypted_in_transit,
        encrypted_at_rest,
        recommendation: recommendation.to_string(),
    }
}

File: services/federation-product-exchange-service/internal/observability/encryption.go
package observability

import (
	"github.com/openfoundry/openfoundry-go/services/federation-product-exchange-service/internal/models"
)

// Posture ports `domain::encryption::posture` 1:1.
//
// Profile defaults to "mutual-tls+envelope" when no contract is provided;
// key_version defaults to "key/pending"; encrypted-* default to true. The
// recommendation is "rotation current" when both encrypted_in_transit and
// encrypted_at_rest are true, otherwise "provision key rotation before
// replication".
func Posture(
	share *models.SharedDataset,
	contract *models.SharingContract,
	syncStatus *models.SyncStatus,
) models.EncryptionPosture {
	profile := "mutual-tls+envelope"
	if contract != nil {
		profile = contract.EncryptionProfile
	}

	keyVersion := "key/pending"
	encryptedInTransit := true
	encryptedAtRest := true
	if syncStatus != nil {
		keyVersion = syncStatus.KeyVersion
		encryptedInTransit = syncStatus.EncryptedInTransit
		encryptedAtRest = syncStatus.EncryptedAtRest
	}

	recommendation := "provision key rotation before replication"
	if encryptedInTransit && encryptedAtRest {
		recommendation = "rotation current"
	}

	transportCipher := "disabled"
	if encryptedInTransit {
		transportCipher = "TLS 1.3 mTLS"
	}
	atRestCipher := "disabled"
	if encryptedAtRest {
		atRestCipher = "AES-256-GCM envelope"
	}

	return models.EncryptionPosture{
		ShareID:            share.ID,
		TransportCipher:    transportCipher,
		AtRestCipher:       atRestCipher,
		KeyVersion:         keyVersion,
		Profile:            profile,
		EncryptedInTransit: encryptedInTransit,
		EncryptedAtRest:    encryptedAtRest,
		Recommendation:     recommendation,
	}
}
